x-pack/plugins/event_log/server/es/cluster_client_adapter.ts

```diff
@@ -30,14 +42,71 @@ export interface QueryEventsBySavedObjectResult {
 export class ClusterClientAdapter {
   private readonly logger: Logger;
   private readonly clusterClientPromise: Promise<EsClusterClient>;
+  private readonly docBuffer$: Subject<Doc>;
+  private readonly docsBufferedSubscription: Subscription;
+  private readonly doneWriting: ReadySignal;
+  private readonly context: EsContext;
 
   constructor(opts: ConstructorOpts) {
     this.logger = opts.logger;
     this.clusterClientPromise = opts.clusterClientPromise;
+    this.context = opts.context;
+    this.docBuffer$ = new Subject<Doc>();
+    this.doneWriting = createReadySignal();
+
+    // buffer event log docs for time / buffer length, ignore empty
+    // buffers, then index the buffered docs
+    const docsBuffered$ = this.docBuffer$.pipe(
+      bufferTime(EVENT_BUFFER_TIME, null, EVENT_BUFFER_LENGTH),
+      filter((docs) => docs.length > 0),
+      switchMap(async (docs) => await this.indexDocuments(docs))
+    );
+
+    // kick everything off, when the stream closes, signal finished
+    const { doneWriting } = this;
+    this.docsBufferedSubscription = docsBuffered$.subscribe({
+      complete() {
+        doneWriting.signal();
+      },
+    });
+  }
+
+  // This will be called at plugin stop() time; the assumption is any plugins
+  // depending on the event_log will already be stopped, and so will not be
+  // writing more event docs.  So we shut down the docBuffer$ observable,
+  // and wait for the docsBufffered$ observable to complete
+  public async shutdown(): Promise<void> {
+    this.docBuffer$.complete();
+    await this.doneWriting.wait();
+    this.docsBufferedSubscription.unsubscribe();
```

Comment:
Line 81 is actually redundant - once the `docBuffer$` completes it will reactively unsubscribe `docsBufferedSubscription` so no need to call this explicitly.